FEATURE: Asynchronous thumbnail generation

aertmann opened the pull request:
Use asynchronous generation of thumbnails to avoid long backend request times when many thumbnails need to be generated.

- Asynchronous thumbnail generation (NEOS-1668)
- Use asynchronous thumbnails (NEOS-1672)
- Media thumbnail presets (NEOS-1669)
- Automatic creation of preset thumbnails (NEOS-1671)
- Command to generate thumbnails from presets (NEOS-1681)
- Clearing by preset in clear thumbnails command (NEOS-1684)
- Command to generate uninitialized thumbnails (NEOS-1670)

Introduces a ``async`` argument to the image view helpers, image server and asset service to create thumbnails without generating the image. A thumbnail controller is introduced which the view helpers
link to when a thumbnail is not generated during rendering. The controller redirects to the thumbnail
when generated.

Asynchronous thumbnails are used in the media browser and asset service by default to improve the
requests times to the backend. It's possible to disable asynchronous thumbnails with the configuration
setting ``TYPO3.Media.asyncThumbnails``.

Since several simultaneous requests for thumbnails can occur, depending on browser
and concurrent users, busy servers can experience problems. Therefore it is recommended to
configure the server to run ``media:generatethumbnails`` often or use a job queue by listening to the
``thumbnailCreated`` signal and calling ``refreshThumbnail`` for the thumbnail in the thumbnail service.

Additionally a concept for thumbnail presets is introduced to allow thumbnails to be reused to reduce
the amount of necessary thumbnails. Thumbnails presets are configured in the configuration setting
``TYPO3.Media.thumbnailPresets``.

When new images are uploaded thumbnails for the configured presets are automatically created, unless disabled using the configuration setting ``TYPO3.Media.autoGenerateThumbnailPresets``.
If asynchronous thumbnails are disabled, the thumbnails will be generated immediately making
uploading slower.

The media browser and the asset services use these presets to reduce the amount of different
thumbnails used and to enable the generating of the thumbnails to be done server side instead
when the user requests them.

Last two new commands ``media:createthumbnails`` and ``media:generatethumbnails`` are introduced for creating by presets and generating uninitialized thumbnails. Support for clearing by preset is also
added to the ``media:clearthumbnails`` command.

NEOS-1667 #close

aertmann commented:
Ready for initial review and feedback very welcome. Works well and turned out nicely IMO.

Missing part is documentation, currently waiting for feedback where to put that since no documentation is available for the package in general.

Conflicts with #21 so will try to work with @dfeyer to find how we solve that in a good way.

aertmann commented:
Have a working copy rebased on top of #21 https://github.com/aertmann/neos-development-collection/tree/media-browser-performance-optimization-rebased

kdambekalns commented on TYPO3.Media/Migrations/Postgresql/Version20151116054040.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function up(Schema $schema) {
+		// this up() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented on TYPO3.Media/Migrations/Postgresql/Version20151116054040.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function down(Schema $schema) {
+		// this down() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented on TYPO3.Media/Migrations/Postgresql/Version20151118143845.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function up(Schema $schema) {
+		// this up() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented on TYPO3.Media/Migrations/Postgresql/Version20151118143845.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function down(Schema $schema) {
+		// this down() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented on TYPO3.Media/Migrations/Mysql/Version20151118143756.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function down(Schema $schema) {
+		// this down() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented on TYPO3.Media/Migrations/Mysql/Version20151118143756.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function up(Schema $schema) {
+		// this up() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented on TYPO3.Media/Migrations/Mysql/Version20151116052338.php:
```diff
@@ -0,0 +1,33 @@
+	 * @return void
+	 */
+	public function down(Schema $schema) {
+		// this down() migration is autogenerated, please modify it to your needs
```
Uhm, you looked at this file? Then get rid of this line!

kdambekalns commented:
Tested this quickly (by checking out this PR and merging master into it). It works… :) 

The only thing I noticed was that some thumbnails stuck with the loading indicator, and appeared only after a reload (this was in the media management). Could be Firefox or whatever, so…

And the media:createthumbnails command exited cleanly but said it had created 34 of 38 thumbnails - yes, no generated thumbnail for the PDF and TTF file in the library (small/large = 4 thumbnails). This is a bit confusing…

dfeyer commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Model/Thumbnail.php:
```diff
@@ -66,29 +68,38 @@ class Thumbnail implements ImageInterface
      * @throws \TYPO3\Media\Exception
      */
-    public function __construct(AssetInterface $originalAsset, ThumbnailConfiguration $configuration)
+    public function __construct(AssetInterface $originalAsset, ThumbnailConfiguration $configuration, $async = false)
```
Why not adding the $async to the ThumbnailConfiguration, I create this object to avoid having tons of properties here ... and look like I did the same things for the $transient property in the Thumbnail Generator who cause conflict with this one. In a perfect world we don't need more properties here, everything should of in the ThumbnailConfiguration

dfeyer commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/AssetService.php:
```diff
@@ -32,71 +35,47 @@ class AssetService
+     * @throws AssetServiceException
      */
-    public function getThumbnailUriAndSizeForAsset(AssetInterface $asset, ThumbnailConfiguration $configuration)
+    public function getThumbnailUriAndSizeForAsset(AssetInterface $asset, ThumbnailConfiguration $configuration, $async = false, ActionRequest $request = null)
```
See previous comment about the ThumbnailConfiguration and the async property, not sure if the ActionRequest should go to the ThumbnailConfiguration

dfeyer commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php:
```diff
@@ -75,23 +80,36 @@ class ThumbnailService
      * @throws \Exception
      */
-    public function getThumbnail(AssetInterface $asset, ThumbnailConfiguration $configuration)
+    public function getThumbnail(AssetInterface $asset, ThumbnailConfiguration $configuration, $async = false)
```
See my previous comment about ThumbnailConfiguration

dfeyer commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Service/ThumbnailService.php:
```diff
@@ -75,23 +80,36 @@ class ThumbnailService
                 throw new NoThumbnailAvailableException(sprintf('ThumbnailService could not generate a thumbnail for asset of type "%s" because currently only Image assets are supported.', get_class($asset)), 1381493670);
             }
-            $thumbnail = new Thumbnail($asset, $configuration);
+            $thumbnail = new Thumbnail($asset, $configuration, $async);
```
See my previous comment about ThumbnailConfiguration

dfeyer commented:
Really nice change, look like there's some conflict with the last version of #21 and we need to move $async, $actionRequest (from this PR) and $transient (from #21) to the ThumbnailConfiguration object, it's fare more manageable to have all those property in a single objet and we simplify the maintenance a lots, changing those method signature is a nightmare. 

It's a bit too late to day to do more, and tomorrow I spend most of the day at customer location. Will have some time betweek 4PM and 5PM, or later in the evening. Ping me if you need me, I check slack during the day.

:+1: from my side, if we move a least $async and $transient to the ThumbnailConfiguration (property are by default transient in the ThumbnailConfiguration, should be added in the toArray method to be stored).

aertmann commented:
:+1: became more complicated on top of the thumbnail generators, but happy with the outcome though.

found a way to get rid of the svg generator which works with the asynchronous thumbnails as well (was causing some issues with static resource linking from CLI)

regarding thumbnails not being generated, tested about 50-75 times today without any issues.. had some earlier due to the ``stat`` issue, but that's unrelated..

also fixed the counters in the commands..

and moved the async flag into the thumbnail configuration..

removed the transient flag from thumbnails and made the static ones persistable since it helps with the asynchronous generation logic and removes creation of links from CLI

documented the whole thing on top and added documentation references for commands, view helpers and validators as well

+ plus additional fixes and tweaks

robertlemke commented on TYPO3.Media/Classes/TYPO3/Media/Command/MediaCommandController.php:
```diff
@@ -109,13 +124,50 @@ public function importResourcesCommand($simulate = false)
 
     /**
+     * Generate thumbnails for thumbnail presets
+     *
```
This lacks documentation for the help screen (further lines in the doc comment) which could explain when you'd want to run this command. Also a reference to clearThumbnails would make sense.

Just a remark: I noticed that you mostly use the term "generate thumbnails" in the code, but the command is called "createThumbnails". Maybe change that to "generateThumbnails" as well?

kdambekalns replied:
It's more complicated: there is a `generateThumbnails` command as well…

robertlemke commented on TYPO3.Media/Classes/TYPO3/Media/Command/MediaCommandController.php:
```diff
@@ -109,13 +124,50 @@ public function importResourcesCommand($simulate = false)
+
+    /**
      * Remove all thumbnail objects and resources
+     *
```
Also add a help message here, and try to keep the first line shorter ("Remove all thumbnails" would suffice, I guess).

kdambekalns commented:
Thanks for fixing my earlier comments. Did you write the command, VH and validator references by hand? I hope not, we autogenerate them… ;)

Ok, will do some testing now…

robertlemke commented on TYPO3.Media/Classes/TYPO3/Media/Command/MediaCommandController.php:
```diff
@@ -123,6 +175,31 @@ public function clearThumbnailsCommand()
+     * @param integer $limit
+     * @return void
+     */
+    public function generateThumbnailsCommand($limit = null)
```
okay, sorry, I'm reading the code from top to bottom, so didn't see that there's another command called "generateThumbnails". I don't know yet about presets, so by now - when I just read the user facing API - it's a bit confusing, what means what.

aertmann replied:
yeah it's been bugging me as well, but haven't thought of something better.. could be ``renderThumbnails`` instead maybe

robertlemke replied:
maybe it's possible to combine both commands into one? I have to admit that I didn't completely get the difference between "thumbnails with and without preset" yet, after skimming all the code. But it's also been a lot of hours today already ...

kdambekalns commented on TYPO3.Media/Classes/TYPO3/Media/Domain/Model/Thumbnail.php:
```diff
@@ -24,9 +24,12 @@
- * 		@ORM\Index(name="originalasset_configurationhash",columns={"originalasset", "configurationhash"})
- *    }
+ *  uniqueConstraints={
+ *      @ORM\UniqueConstraint(name="originalasset_configurationhash",columns={"originalasset", "configurationhash"})
```
The name for this index (and the next one a few lines down) are not reflected in the DB migrations.

robertlemke commented on TYPO3.Media/Documentation/Commands.rst:
```diff
@@ -0,0 +1,25 @@
+
+By default it creates for all configured presets, unless a specific preset is added using the ``preset`` parameter.
+
+Asynchronous thumbnails are
```
Asynchronous thumbnails are … ?

kdambekalns replied:
If they weren't, we'd be in real trouble!

But, good question. Are they?

aertmann replied:
no.

kdambekalns commented on TYPO3.Media/Documentation/AsynchronousThumbnails.rst:
```diff
@@ -0,0 +1,41 @@
+
+To optimize response times, generation of thumbnails can be done asynchronously. Usage of asynchronous thumbnail
+generation is determined in the image view helpers usage with the ``async`` flag. When the flag is used, a link to the
+thumbnail controller returned instead of the generating the thumbnail if the thumbnails hasn't already been generated.
```
instead of generating the thumbnail if the thumbnail hasn't already been generated

kdambekalns commented on TYPO3.Media/Documentation/AsynchronousThumbnails.rst:
```diff
@@ -0,0 +1,41 @@
+The configuration setting ``TYPO3.Media.asyncThumbnails`` is used to determine if asynchronous thumbnails are generated
+when creating thumbnails for configured :ref:`Thumbnails Presets`.
+
+The setting is additionally used as the default value for the ``media:createthumbnails``,
```
the ``media:createthumbnails`` command

kdambekalns commented:
Left some comments and would like to see the index names used in the DB migrations.

And thanks for the documentation!

robertlemke commented:
Some thoughts after reading the code:

- wow, what a big change and what a nice feature ;-)

- it should get clearer for administrators what the different commands do and what thumbnail presets are. I think that the command documentation (within the command controller) would be a good place for that kind of documentation

- from how I interpret the code, it would be possible to flood the server with requests which would produce large amounts of thumbnails if I know an `Asset` uuid. Did you consider that? If that would be the case, I'd consider this a security issue we have to deal with before we could merge it.

kdambekalns commented:
The images not being generated for my turned out to be 500 errors caused by that dreaded locking issue… So this change works fine, as it seems.

Regarding documentation, yes, this could surely be improved, but if we are talking about merge-critical stuff I'd say it's

- the index names in the DB migration
- the flooding @robertlemke mentioned

If that is solved, it can go in from my side.

robertlemke commented:
Yes, of course the documentation isn't critical for merging it tonight. But regarding the potential security issue, I think it's a bit risky to merge the change and then try to make it secure, because this might need some additional mechanisms which I can't oversee yet. If anonymous requests should be able to trigger thumbnail creation via the `thumbnailAction()`, we must provide some layer of security which protects the server against misuse of the feature, for example by signing the thumbnail configuration.

kitsunet commented:
To document discussion from Slack: Security won't be an issue due to the PropertyMapper preventing Object creation

robertlemke commented:
@kitsunet just to be sure, I tried a few ways to trigger thumbnail creation through modified request, but this does seem to be secure.

robertlemke commented:
Verified the feature from a user's point of view, skimmed all of the code, but not very thoroughly.
Can't test the fixed migration right now, but if that is working well for migrating from 2.0 to 2.1 then this change should be fine to be merged.

:+1:

kitsunet commented:
:+1:

kdambekalns commented:
:+1:

kdambekalns commented:
I'd say merge, as soon as Travis turns green. :tada:

Thanks for the patience @aertmann and to everyone reviewing for the endurance!

aertmann commented:
:+1: thanks a lot for helping reviewing this beast guys! much appreciated :)

aertmann commented:
lift off :rocket:
